### ISSUE TITLE
PFX secret are not converted to PEM

### ISSUE THREAD

**Issue by deuch** [Azure/secrets-store-csi-driver-provider-azure#273]
**What steps did you take and what happened:**

I’ve created a certificate with App Service Certificate. This one is stored as a secret in a keyvault (PFX with cert and key embedded) in base64. It’s done like that by Azure.

I’ve tried your sample for ingress TLS certificate and have issue.
The certificate is reprieves from Keyvault but not converted in PEM in my container.
I’ve tried objectFormat: pem or pfx without luck.
I’ve played with obectEncoding too (base64) and the main difference is that the pfx is in base64 or binary in my pod.

The secret is created but with tls.crt to empty and tls.key with a base64 encoded string with only the BEGIN PRIVATE KEY and END PRIVATE KEY markers, not the value.

I’m using AKS 1.18 and the latest version of csi driver (helm chart 0.0.13)

 No error in logs ... in debug mode

Do you have a sample with PFX secret as source certificate from Keyvault to PEM and K8s secret target ?

**What did you expect to happen:**

Certificate as PEM in my container.
Kubernetes tls secret with tls.crt and tls.key populated with the right values
.
**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Which access mode did you use to access the Azure Key Vault instance:**
 Pod Identity

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): 0.0.16
- Azure Key Vault provider version: (use the image tag): 0.0.9
- Kubernetes version: (use `kubectl version` and `kubectl get nodes -o wide`): 1.18.8
- Cluster type: (e.g. AKS, aks-engine, etc): AKS

**Comment by mortenlerudjordet** [comment]
I'm on the same versions.

What works for me in SecretProviderClass:
```
  objectName: "NameOfPFXcertInKV"
  objectAlias: "certificate.pem" #this will be the name of the file exported to the container
  objectType: "secret"
  objectFormat: "pem"
```
For me this gives me both the private and public key in the pem file set in objectAlias
I tested this by using **kubectl cp** to get the cert file from a pod that has a secret store mountpoint active.

If you are not getting what you expect I would download the pfx from the keyvault and use openssl to get content:
```
openssl pkcs12 -in kvfile.pfx -out cert.pem -passin pass:"" -nodes
```
Not pfx files dl from kv has pwd set to empty string. Then open it up and check that the keys are there.

**Comment by deuch** [comment]
Tested your conf and still have the base64 encoded pfx in the container.

The certificate came from app service certificate and is created by azure.
I do not want to modify it because I will loose the auto-renew feature.

I’ve convert the certificate in PEM and convert it to pfx without luck.
The public and private key are in the same cert and store as a secret. In my KV there is no key or certificate associated to this PFX. Only one secret with a pfx (cert+key with full goddaddy certification chain)

**Comment by mortenlerudjordet** [comment]
In my kv the pfx is stored as a certificate. So if you have stored it as a secret, this might be why you have a different experience.
I also use the pfx in another container, but before using it there I have to do a initContiner using tannerfe/alpine-openssl image and do a command:
```
[
  "sh",
  "-c",
  "cat /secrets/pfxbase64 | base64 -d - > /ssl/certificate.pfx",
]
```
With the following in SecretProviderClass:
```
- |
  objectName: "NameOfPFXcertInKV"
  objectAlias: "pfxbase64"
  objectType: "secret"
  objectFormat: "pfx"
  objectVersion: ""
```

**Comment by deuch** [comment]
Yes we already do that for converting pfx to pem with initContainer.

My aim is to have this secret injected in a k8s tls secret and set the auto-rotation of csi-driver. This is a new feature released a few day ago I’m waiting for ...
With the 2 auto-rotation (auto-renew of certificate in KV and auto-rotation in csi drivers) the ingress tls certificate are always up to date ...

**Comment by aramase** [comment]
@deuch The certificate in PFX/PEM format needs to be stored as a certificate. Key Vault on the backend will split the components and store them as secret, key and cert. If the PFX data is stored as a secret it's just a blob which will be returned as is by the Azure Key Vault provider or `az keyvault secret download`. 

If using the CLI, you can upload the PFX file by running `az keyvault certificate import`. Ref - https://docs.microsoft.com/en-us/azure/key-vault/certificates/tutorial-import-certificate

You can then follow the doc here to determine what object type to use for CSI driver - https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/docs/getting-certs-and-keys.md

**Comment by deuch** [comment]
I do not import the certificate in keyvault by myself, it’s the azure managed service App Service Cerificate that do it. When t’yoinbuy a certificate from App Service Certificate, you need to choose (mandatory) a keyvault to store it. And it is store as a secret, n’ont a certificate.

If I export and import it as certificate and key I will list the auto-renew and rekey features of app service certificate. Those both features are very important for us.

**Comment by aramase** [comment]
Thank you for the details @deuch!

Looks like upload the certificate directly as a secret is deprecated and the recommended way is to import it as a certificate. 

![Screen Shot 2020-10-16 at 4 12 44 PM](https://user-images.githubusercontent.com/10608509/96321518-9571b680-0fca-11eb-9646-ad79f880a8ce.png)

@jlichwa Do you know when this option will be removed? Should App service be storing the pfx data as a certificate instead?

**Comment by deuch** [comment]
> Thank you for the details @deuch!
> 
> Looks like upload the certificate directly as a secret is deprecated and the recommended way is to import it as a certificate.
> 
> ![Screen Shot 2020-10-16 at 4 12 44 PM](https://user-images.githubusercontent.com/10608509/96321518-9571b680-0fca-11eb-9646-ad79f880a8ce.png)
> 
> @jlichwa Do you know when this option will be removed? Should App service be storing the pfx data as a certificate instead?

**Comment by deuch** [comment]
@aramase Maybe it’s not the official way to import certificate, but it’s how app service certificate work right now :) Only Azure can change this behavior. And it will be nice to have it ASAP to avoid issue in production with expired certificate. 

We have everything available in Azure to auto-rotate the cerificate but the azure components are not compatible between them🤨

Again, having this working (auto-rotate of app service certificate in AKS) is truly a killing feature ! 🤯

**Comment by jlichwa** [comment]
@deuch I don't think that is how it is how it is recommended for managing app service certificate. You should not upload it as secret. Some guidelines here https://docs.microsoft.com/en-us/azure/app-service/configure-ssl-certificate#import-a-certificate-from-key-vault also adding @sebansal

**Comment by deuch** [comment]
App service certificate is a managed service of Azure that generate certificates from GoDaddy. It’s app service certificate that store the certificate as secret furring creation not me. 

You have no choice to do it in another way. It’s fully done by Azure. I just set the domain and the keyvault and all is automatically store in keyvault as secret and not certificate. 

If the official service to generate certificate does not work well with keyvault, it’s not my fault 😀 and I would like to have a solution for that. To be honest I do not care if the certificate is store as secret or certificate, I just want to have a fully functional
Integration between 3 azure/Microsoft product (AKS/CSI/KV/App Service Certificate)

@aramase do you have an answer for this strange behavior ?

**Comment by jlichwa** [comment]
@deuch I'm representing key vault, and this CSI Driver for Key Vault. If AppService does not follow our best practices, that is not our fault. You can create issue on them. #please-close

**Comment by sebansal** [comment]
An App Service Certificate is stored as a Key Vault secret. You can export a PFX copy of it and use it for other Azure services or outside of Azure.
How-to commands are provided here by the App Service Team.
https://docs.microsoft.com/en-us/azure/app-service/configure-ssl-certificate#export-certificate

**Comment by sebansal** [comment]
The downloaded appservicecertificate.pfx file is a raw PKCS12 file that contains both the public and private certificates.

**Comment by deuch** [comment]
> An App Service Certificate is stored as a Key Vault secret. You can export a PFX copy of it and use it for other Azure services or outside of Azure.
> How-to commands are provided here by the App Service Team.
> https://docs.microsoft.com/en-us/azure/app-service/configure-ssl-certificate#export-certificate

Doing that and I lost the auro-renew feature because I need to develop a synchronisation tool between the certificate as secret and the one I exported ... many many productions incident to come ...

Everything is already setup with app service certificate. We just need to be able to retrieve a pfx stored a secret with the csi driver, to mount it as PEM files or PFX in the pod and create the according k8s secret from the PEM files (for ingress purpose).

The certificate is renew in the keyvault by app service ? No issue, with the new auto-rotation feature, new certificate will be retrieved from keyvault and updated in the pod and in the k8s secret ?

Does it make sense to have this feature to avoid update manually certificate in our app ? With auto-rotation set to 24h between check, and the auto-renew 60 days before the expiration date, no more certificate expiration issue. We deploy at least 1 time per week, so pod have always the last version of the certificate.
Our ingress controller (nginx) watch secrets it used and updated its configuration when they changed.

If I put something in the middle or do an import/export, I lost everything that I explained.

**Comment by deuch** [comment]
> @deuch I'm representing key vault, and this CSI Driver for Key Vault. If AppService does not follow our best practices, that is not our fault. You can create issue on them, they need to update their process to follow best practices. For AppService I'm sure they are already notified and they will have to update their process or their ability to upload it to Key Vault will stop working - I will follow up, once I know the timeline (I expect that to still be working next year).
> 
> We also added PFX to PEM conversion feature to our backlog. We will assign priority and consider it for next year planning.

You already have the pfx to PEM. You just have to call the method when using a secret instead of a certificate (decodePKCS12 function in provider.go). But you will have limitation with the design of provider.go.

**Comment by sebansal** [comment]
As @aramase mentioned, the app service cert needs to be stored as a cert. Perhaps Azure app service certificate team may be able to assist further. 
@Azure/appservice

**Comment by kikokikok** [comment]
Or you could try to talk within your organization Azure team A to Azure team B on the things to be aligned to make Azure as a whole more consistent and easy to use from a client perspective.
Anyway we will have the topic escalated through the CSAs where it makes most sense to have it considered.

Thanks anyway !

**Comment by ritazh** [comment]
@deuch I share your frustration with this. Let us find the right contact in Azure App Service and get back to you on this issue.

**Comment by deuch** [comment]
> @deuch I share your frustration with this. Let us find the right contact in Azure App Service and get back to you on this issue.

Thank you !

**Comment by stefsch** [comment]
There is a long term plan to change the object type to "certificate" for App Service Certificates (ASC) that are generated and stored in KeyVault.  For the near term though, it is an implementation limitation of the feature that it stores certificates as type "secret".

**Comment by deuch** [comment]
In the mean time can we have a patch in the csi driver to be able to use ASC certificates as secret ?

**Comment by alarco-l** [comment]
Is there any news on a potential patch in the csi driver ?
On an user perspective, Azure is an all in one service, we expect services to work nicely all together.
I understand it's a **long term plan to change the object type to "certificate" for App Service Certificates (ASC)** but in the meanwhile we have production outage because of that.
My team and I will really appreciate a patch in the csi driver as of today the only solution seems weird. Export the certificate and re-import it, will break the auto renew, which is the point to use ASC.

**Comment by aramase** [comment]
@stefsch Gentle ping! Are there any updates for App Service Certificates?

**Comment by deuch** [comment]
> @stefsch Gentle ping! Are there any updates for App Service Certificates?

Same here !

**Comment by kyschouv** [comment]
Pinging to keep this alive, as I'm hoping there's a resolution soon. We have a hack in place in the meantime.

**Comment by sebansal** [comment]
Adding @shisriva-msft

**Comment by thcp** [comment]
Currently facing the same issue here, pinging to keep this issue alive.

**Comment by deuch** [comment]
They made it …
https://akv2k8s.io/tutorials/sync/9-certificate-as-AKV-secret/

**Comment by thcp** [comment]
Commenting to keep this issue alive.

**Comment by deuch** [comment]
Almost 1 year and no news ?

**Comment by yutanglin16** [comment]
This work to change App Service Certificate as a "certificate" type in KV is still planned and is on our backlog . We don't have any ETA to provide yet regarding this. I am not expecting us to publicly support it this calendar year.

**Comment by carlo-quinonez** [comment]
Update on this because I just ran into this as well.

1. When an AppService Certificate (ASC) is purchased, the PFX cert is stored in AKS as a Secret.
2. When you add the certificate to an AppService, a new AKV Certificate is created in KeyVault. It appears the AKV Certificate is just a copy of the original AKV Secret.

Clearly, managing the AKS Secrets -> Certificate conversion has already been implemented. Maybe AKS could leverage that same mechanism?

**Comment by alessio-trivisonno-shift** [comment]
2 years and still not solved... unbelievable people 🤯

**Comment by joaop221** [comment]
this issue will turn **3 years** 🎉

**Comment by ACavuoto** [comment]
I spent the last 2 days trying to figure out why I couldn't make TLS to work, and finally ended up on this page by chance.

It's inconceivable that two teams in the same company keep blaming each other about who is not doing what right, but nobody had the decency of hot-fixing the issue or, at least, documenting it in their own "how to" guide.

This is deeply disappointing.

**Comment by somjex** [comment]
Still haven't found how to work around this issue ...

**Comment by carlo-quinonez** [comment]
@somjex
1. Purchase AppService Certificate (ASC) and store in Azure Key Vault (AKV) 
2. Create a dummy AppService
3. Add custom domain on AppService and use the ASC that's stored as AKV Secret. 
5. Et Voila, there's a new Certificate item in AKV that works with AKS..

**Comment by d3st1nyh4x** [comment]
provided workaround does not work, at least not with wild card certificates.

I created a custom domain, 
select import certificate later, 
import the certificate from existing ASC, 

there is no new certificate item in AKV, but the cert was added to the app service...

coming up on 4 years since this ticket was opened, 
the conversion from PFX to PEM is trivial, how is this still an issue?

**Comment by darreng1234** [comment]
Commenting to keep the issue alive. We need to export the ASC as a certificate object to the KV or else the E2E automation breaks with CSI driver.
